File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.py
import heapq
n, e = map(int, input().split())
edges = [[] for _ in range(n+1)]
for _ in range(e):
    a, b, cost = map(int, input().split())
    edges[a].append((b, cost))
    edges[b].append((a, cost))
v1, v2 = map(int, input().split())
INF = int(1e9)

def djikstra(start):
    dist = [INF] * (n+1)
    dist[start] = 0
    pq = [(0, start)]
    
    while pq:
        cost, node = heapq.heappop(pq)
        
        if cost > dist[node]:
            continue
        
        for nxt, nc in edges[node]:
            new_cost = nc + cost
            if new_cost < dist[nxt]:
                dist[nxt] = new_cost
                heapq.heappush(pq, (new_cost, nxt))
    return dist
d_start = djikstra(1)
d_v1 = djikstra(v1)
d_v2 = djikstra(v2)

p1 = d_start[v1] + d_v1[v2] + d_v2[n]
p2 = d_start[v2] + d_v2[v1] + d_v1[n]
ans = min(p1,p2)

print(ans if ans < INF else -1)

File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Scanner;


//TIP To <b>Run</b> code, press <shortcut actionId="Run"/> or
// click the <icon src="AllIcons.Actions.Execute"/> icon in the gutter.
public class Main {

	static class Node {
		int vertex;
		int weight;

		Node(int vertex, int weight) {
			this. vertex = vertex;
			this.weight = weight;
		}
	}

	static int[] djikstra(int start) {
		int[] route = new int[n+1];
		PriorityQueue<Node> pq = new PriorityQueue<>(Comparator.comparingInt(n -> n.weight));
		pq.add(new Node(start, 0));
		Arrays.fill(route, INF);
		route[start] = 0;

		while (!pq.isEmpty()) {
			Node node = pq.poll();
			int totW = node.weight;
			int curV = node.vertex;

			if (totW > route[curV]) continue;

			for (Node n : graph[curV]) {
				int newDist = totW + n.weight;
				if(newDist < route[n.vertex]) {
					route[n.vertex] = newDist;
					pq.add(new Node(n.vertex, newDist));
				}
			}
		}
		return route;
	}

	static final Integer INF = Integer.MAX_VALUE;
	static int n, e;
	static List<Node>[] graph;

	public static void main(String[] args) {
		long INF_L = INF;

		Scanner sc = new Scanner(System.in);
		n = sc.nextInt();
		e = sc.nextInt();
		graph = new ArrayList[n+1];
		for(int i = 0; i <= n; i++) {
			graph[i] = new ArrayList<>();
		}


		for(int i = 0; i < e; i++) {
			int u = sc.nextInt();
			int v = sc.nextInt();
			int w = sc.nextInt();
			graph[u].add(new Node(v, w));
			graph[v].add(new Node(u, w));
		}
		int v1 = sc.nextInt();
		int v2 = sc.nextInt();

		int[] distFromOne = djikstra(1);
		int[] distFromV1 = djikstra(v1);
		int[] distFromV2 = djikstra(v2);

		long r1 = (long)distFromOne[v1] + distFromV1[v2] + distFromV2[n];
		long r2 = (long)distFromOne[v2] + distFromV2[v1] + distFromV1[n];
		long ans = Math.min(r1, r2);

		System.out.println(ans >= INF_L ? -1 : ans);

		sc.close();
	}
}